src/ring_theory/localization/away.lean

```diff
@@ -194,3 +196,69 @@ lemma is_localization.away.finite_presentation (r : R) {S} [comm_ring S] [algebr
   [is_localization.away r S] : algebra.finite_presentation R S :=
 (adjoin_root.finite_presentation _).equiv $ (localization.away_equiv_adjoin r).symm.trans $
   is_localization.alg_equiv (submonoid.powers r) _ _
+
+section num_denom
+
+open multiplicity unique_factorization_monoid is_localization
+
+variable (x : R)
+
+variables (B : Type*) [comm_ring B] [algebra R B] [is_localization.away x B]
+variable (hx : irreducible x)
+
+/-- `self_as_unit` is the element `x` at which `B` is a localization away, bundled as a unit in `B`
+-/
+noncomputable def self_as_unit : Bˣ := ⟨algebra_map _ _ x, away.inv_self x, away.mul_inv_self _,
```

Comment:
I must say that I don't get the motivation for this definition yet: it seems to always be cast back to `B`. Could you explain (in the docstring) what this allows us to do more easily than with `algebra_map _ _ x`?

Reply:
I see. Maybe an alternative that avoids the coercions would be to define something like `self_zpow : B`, which is `x ^ n` if `n` is nonnegative and `mk 1 x^(-n)` if `n` is negative. I'm happy to try it out for you, just don't have much time today.

Reply:
OK, sure, but why do you think that my way is inappropriate? Is it "simply" ugly or do you see potential harms?

Reply:
I wouldn't say inappropriate is precisely the right word but I can't think of a better one so let's go with yours :)

The reason I complain is not so much a concretely identifiable harm, more of an intuitive warning. Namely, if we do end up with a definition that needs a relatively high amount of ceremony to use, *especially* one that is rarely used, maintainers* are going to have to figure out how to deal with this each time they pass through this area of the codebase. Certainly this will be relevant for the mathlib4 port. (Especially if you or I can't help out the maintainer to deal with understanding why things work the way they do.) And changing the definition later on is going to cause a lot more headaches. So spending some time getting to an elegant definition, or at least an explanation why something may appear less elegant, will save us time and frustration in the end.

What specifically caused my intuition to warn me was that the PR introduces this new notion of `x` as a unit, but then uses it in one specific way (raising it to an integer power and immediately casting away the unit) that hides somewhat the property of being a unit. Moreover, these nested type ascriptions `(((self_as_unit x B)^n : Bˣ) : B)` at the use site might be hard to get right if you don't know Lean's elaboration strategy, and the fact they are repeated at each mention of `self_as_unit` suggests the repetitive part, that is likely to be used wrong, might be incorporated into the definition itself, hence `self_zpow`.

Reply:
I see, it certainly makes sense. For the time being I have moved the lemma to UFD and opened the corresponding `mathlib4 ` PR. I will think about your suggestion and see if I can come up with a better idea.

Reply:
@Vierkantor I have `pushed` a tentative definition with some lemmas. Is this the direction you were suggesting?

Reply:
That is indeed what I meant. Let me try and solve those `sorry`s...

Reply:
I went ahead and pushed a slightly different definition, using `int.nat_abs` to convert integers to natural numbers.

Reply:
I picked up on your work and closed some ` sorry` s, only half of a lemma seems left.